File: src/research_core/jurisdiction_skills.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
import re
import unicodedata


JURISDICTIONS_ROOT = (
    Path(__file__).resolve().parents[1] / "lib" / "research" / "skills" / "jurisdictions"
)


@dataclass(frozen=True)
class JurisdictionSkill:
    id: str
    content: str


@dataclass(frozen=True)
class JurisdictionSkillsResolution:
    county: JurisdictionSkill | None
    city: JurisdictionSkill | None
    gaps: tuple[str, ...]


def _deaccent(value: str) -> str:
    normalized = unicodedata.normalize("NFD", value)
    return "".join(ch for ch in normalized if unicodedata.category(ch) != "Mn")


def normalize_county_name(county: str) -> str:
    return re.sub(r"\s+county\s*$", "", county.strip(), flags=re.IGNORECASE)


def _slug_part(value: str) -> str:
    deaccented = _deaccent(value).strip().lower()
    return re.sub(r"[^a-z0-9]+", "-", deaccented).strip("-")


def _slug_county(county: str) -> str:
    return f"{_slug_part(normalize_county_name(county))}-county"


def _slug_city(city: str) -> str:
    return f"city-of-{_slug_part(city)}"


def jurisdiction_skill_id(county: str, city: str | None = None) -> str:
    county_id = _slug_county(county)
    if city:
        return f"{county_id}/{_slug_city(city)}"
    return county_id


def _read_skill(skill_id: str) -> JurisdictionSkill | None:
    path = JURISDICTIONS_ROOT / skill_id / "JURISDICTION.md"
    if not path.exists():
        return None
    return JurisdictionSkill(id=skill_id, content=path.read_text(encoding="utf-8"))


def resolve_jurisdiction_skills(
    loc: dict[str, str | None],
) -> JurisdictionSkillsResolution:
    county_value = loc["county"]
    county_id = _slug_county(county_value)
    county = _read_skill(county_id)
    gaps: list[str] = []
    if county is None:
        gaps.append(f"county:{county_id}")

    city: JurisdictionSkill | None = None
    city_value = loc.get("city")
    if city_value:
        city_id = f"{county_id}/{_slug_city(city_value)}"
        city = _read_skill(city_id)
        if city is None:
            gaps.append(f"city:{city_id}")

    return JurisdictionSkillsResolution(county=county, city=city, gaps=tuple(gaps))

File: src/lib/research/jurisdictionSkills.ts
import { readFileSync, existsSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

// The county/city jurisdiction skill tree makes the (otherwise unbounded) local
// layer MANAGEABLE by structure: each jurisdiction's local EHS authorities (CUPA,
// fire AHJ, building dept, code amendments) live in their own folder under
// src/lib/research/skills/jurisdictions/<county>/[<city>/]JURISDICTION.md.
//
// The resolver returns whatever folders EXIST for a location and reports the
// rest as explicit gaps — never a guess. Researching a new city = adding its
// folder; the resolver picks it up with no code change.

const HERE = dirname(fileURLToPath(import.meta.url));
const ROOT = join(HERE, "skills", "jurisdictions");

// Strip diacritics so "San José" and "San Jose" both map to the same folder id.
function deaccent(s: string): string {
  return s.normalize("NFD").replace(/[̀-ͯ]/g, "");
}
function slugCounty(county: string): string {
  return deaccent(county).trim().toLowerCase().replace(/\s+/g, "-").replace(/-county$/, "") + "-county";
}
function slugCity(city: string): string {
  return "city-of-" + deaccent(city).trim().toLowerCase().replace(/\s+/g, "-");
}

/** The folder id for a county (and optional city) — the path under jurisdictions/. */
export function jurisdictionSkillId(county: string, city?: string): string {
  const c = slugCounty(county);
  return city ? `${c}/${slugCity(city)}` : c;
}

export type JurisdictionSkill = { id: string; content: string };
export type JurisdictionResolution = {
  county: JurisdictionSkill | null;
  city: JurisdictionSkill | null;
  // Levels expected for this location that have no folder yet (honest gaps).
  gaps: string[];
};

function read(id: string): JurisdictionSkill | null {
  const path = join(ROOT, id, "JURISDICTION.md");
  if (!existsSync(path)) return null;
  return { id, content: readFileSync(path, "utf8") };
}

/** Walk the county/city tree for a location. Returns the skills that exist and
 *  lists the levels that are not yet researched as gaps. */
export function resolveJurisdictionSkills(loc: { county: string; city?: string }): JurisdictionResolution {
  const countyId = slugCounty(loc.county);
  const county = read(countyId);
  const gaps: string[] = [];
  if (!county) gaps.push(`county:${countyId}`);

  let city: JurisdictionSkill | null = null;
  if (loc.city) {
    const cityId = `${countyId}/${slugCity(loc.city)}`;
    city = read(cityId);
    if (!city) gaps.push(`city:${cityId}`);
  }
  return { county, city, gaps };
}
